exercises/zebra-puzzle/example.go

```diff
@@ -0,0 +1,239 @@
+package zebra_puzzle
+
+import "fmt"
+
+/* The Zebra Puzzle facts given:
+
+1. There are five houses.
+2. The Englishman lives in the red house.
+3. The Spaniard owns the dog.
+4. Coffee is drunk in the green house.
+5. The Ukrainian drinks tea.
+6. The green house is immediately to the right of the ivory house.
+7. The Old Gold smoker owns snails.
+8. Kools are smoked in the yellow house.
+9. Milk is drunk in the middle house.
+10. The Norwegian lives in the first house.
+11. The man who smokes Chesterfields lives in the house next to the man with the fox.
+12. Kools are smoked in the house next to the house where the horse is kept.
+13. The Lucky Strike smoker drinks orange juice.
+14. The Japanese smokes Parliaments.
+15. The Norwegian lives next to the blue house.
+
+*/
+
+// Solve returns the answers to the two questions in the zebra puzzle,
+// which are "Who drinks water?", and "Who owns the zebra?"
+func Solve() (waterDrinker string, zebraOwner string) {
+
+	// This solution is modeled somewhat after the Exercism-Python solution.
+
+	// 1. There are five houses.
+	const (
+		firstHouse = iota // name/identity of each house
+		secondHouse
+		middleHouse
+		fourthHouse
+		fifthHouse
+		numHouses
+	)
+	houses := []int{firstHouse, secondHouse, middleHouse, fourthHouse, fifthHouse}
+
+	// Generate the permutations of all the house identities.
+	houseIdentityPermutations := permutations(houses, numHouses)
+
+	// Note: Below, the pattern of var naming of "iXyz" is notation for "potential house (i)dentity of Xyz"
+	// So for example, when comparing pEnglishman == pRed, the test is determining whether the permutation
+	// of placements for both the Englishman and the red house agree to meet the given fact #2.
+
+	for _, houseColors := range houseIdentityPermutations {
+
+		iRed, iGreen, iIvory, iYellow, iBlue := assign(houseColors)
+
+		// 6. The green house is immediately to the right of the ivory house.
+		if justRightOf(iGreen, iIvory) {
+
+			for _, residents := range houseIdentityPermutations {
+
+				iEnglishman, iSpaniard, iUkranian, iJapanese, iNorwegian := assign(residents)
+
+				// 2. The Englishman lives in the red house.
+				// 10. The Norwegian lives in the first house.
+				// 15. The Norwegian lives next to the blue house.
+				if iEnglishman == iRed && iNorwegian == firstHouse && nextTo(iNorwegian, iBlue) {
```

Comment:
I think reversing logic and doing continue might help deep nesting.

Reply:
Hmm.  I'll work with that some.   I was a little unhappy with the deep nesting myself.